### ISSUE TITLE
Adversarial Contrastive Estimation

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#23]
### Metadata
- Authors: Avishek Joey Bose, huan ling, Yanshuai Cao
- Organization: Borealis AI & University of Toronto
- Conference: ACL 2018
- Paper: http://aclweb.org/anthology/P18-1094

**Comment by howardyclo** [comment]
### Metadata
- Authors: Avishek Joey Bose, huan ling, Yanshuai Cao
- Organization: Borealis AI & University of Toronto
- Conference: ACL 2018
- Paper: http://aclweb.org/anthology/P18-1094

### Tags
word embeddings, knowledge graph embeddings, noise contrastive estimation, adversarial training, language modeling (#19), online hard example mining,

### Summary
This paper proposes to augment the negative sampling process in contrastive learning with an *adversarially* learned conditional distribution (conditioned on positive examples), resulting in a negative sampler that adapts to the data distribution and training dynamics and finds harder negatives (refered to as *Adversarial Contrastive Estimation* (ACE)). They demonstrate the efficacy and generality of ACE on learning word embeddings, order embeddings and knowledge graph embeddings.

### Motivation to Use Contrastive Learning
- Computational efficiency, especially when the vocabulary is large.
- Modeling need: For example, modeling entity relations in a vector space naturally leads to a distance-based score to be minimized for observed *entity-relation-entity* triplets. 

### Motivation of This Paper (Problems of Previous Approach)
- Current negative sampling process does not condition on positive example (random sampling). 
- Furthermore, it usually encodes a human prior as to what constitutes a hard negative example, rather than being learned from data. 
- Easy negative examples are sub-optimal for learning discriminative representation as they do not force the model to find critical characteristics of observed positive examples.
- Even if hard negatives are occasionally reached, the infrequency leads to slow convergence.

### Background: Contrastive Learning
Contrastive learning is a sampling-based learning method that contrasting losses on observed positive examples with those on sampled fictitious negative examples, trying to minimize loss on positive ones while maximizing it on negative ones. One of the popular algorithms is *Noise Contrastive Estimation* (NCE) and has been applied on learning word embeddings.

### Purposed Method: Adversarial Contrastive Estimation (ACE)
- Just like *Generative Adversarial Networks* (GAN), ACE consists a discriminator (D) that tries to distinguish between a real positive example and a negative sample selected by **the mixture of a fixed NCE sampler and an adversarial generator** (G).

#### Objectives for D and G: Adversarial mixture noise
> TODO

#### Regularization for G: Entropy and Training
> TODO

#### Tricks: Handling false negatives
ACE samples false negatives more than NCE. Thus, two strategies are further introduced (although entropy regularization reduces this problem):
- Detect if a negative sample is an actual positive observation, make it zero loss for D, and the reward for false negative samples are replace by a large penalty for G.
- Prevent null computation when G learns to sample false negatives, which are subsequently ignored by the D.

### Recommended Readings
- [Notes on Noise Contrastive Estimation and Negative Sampling](https://arxiv.org/pdf/1410.8251.pdf) by Chris Dyer.
- [On word embeddings - Part 2: Approximating the Softmax](http://ruder.io/word-embeddings-softmax/) by Sebastian Ruder.

**Comment by howardyclo** [comment]
### Notes on Noise Contrastive Estimation (NCE)

#### Idea
In neural language modeling (i.e., learn a model distribution *pθ(w | c)* = *uθ(w, c) / Z(c)* to approximate the empirical distribution *p\*(w|c)* as closely as possible.), computing the probability normalization term *Z(c)* in softmax is expensive. Thus, NCE and its simplified variants, *negative sampling* transform the computationally expensive learning problem into a binary classification *proxy problem* that use the same parameters θ to distinguish samples from the empirical distribution (i.e., *(w+, c)~p\*(w|c)*) from samples generated the noise distribution *q(w)* (i.e., (w-, c)~q(w)). In practice, *q(w)* is a uniform unigram distribution). Note: we denote *w+* is a positive example, *w-* is a negative example and *c* is a given condition.

#### Training
During training, we form the training data by sampling one positive example (w+, c) and *k* negative examples (w-, c) with labels 1 and 0 respectively.

Formulate the above description into a mathematical expression:

![NCE1](https://github.com/howardyclo/papernotes/blob/master/images/Notes%20on%20Noise%20Contrastive%20Estimation%20and%20Negative%20Sampling/NCE1.png?raw=true)

And using the definition of conditional probability:

![NCE2](https://github.com/howardyclo/papernotes/blob/master/images/Notes%20on%20Noise%20Contrastive%20Estimation%20and%20Negative%20Sampling/NCE2.png?raw=true)

NCE replaces the empirical distribution *p\*(w|c)* with the model distribution *pθ(w|c)* and makes 2 assumptions to solve the computation problem in *Z(c)*:
- For every *c*, use a parameter *z_c* to estimate *Z(c)*.
- For neural networks with lots of parameters, it turns out that fixing *z_c* = 1 for all *c* is effective. This assumption reduces *z_c* and encourages the model to have "self-normalized* outputs (i.e., *Z(c)* ≈ 1).

Making these assumptions, rewrite the *p\*(w|c)* (which is replaced with *pθ(w|c)*) to *uθ(w, c)*:

![NCE5](https://github.com/howardyclo/papernotes/blob/master/images/Notes%20on%20Noise%20Contrastive%20Estimation%20and%20Negative%20Sampling/NCE3.png?raw=true)

The objective function of maximizing the above conditional log-likelihood of *D* in above equations would be:

![NCE3](https://github.com/howardyclo/papernotes/blob/master/images/Notes%20on%20Noise%20Contrastive%20Estimation%20and%20Negative%20Sampling/NCE3.png?raw=true)

Since the expectation of the second term is still a difficult summation, we use Monte Carlo sampling to approximate the expectation term:

![NCE4](https://github.com/howardyclo/papernotes/blob/master/images/Notes%20on%20Noise%20Contrastive%20Estimation%20and%20Negative%20Sampling/NCE4.png?raw=true)
